dll/win32/kernel32/kernel32_vista/vista.c

```diff
@@ -58,22 +99,78 @@ QueryFullProcessImageNameW(HANDLE hProcess,
                                            &Needed);
         Result = DynamicBuffer;
     }
-    else Result = (PUNICODE_STRING)Buffer;
-
-    if (!NT_SUCCESS(Status)) goto Cleanup;
+    else
+    {
+        Result = (PUNICODE_STRING)Buffer;
+    }
 
-    if (Result->Length / sizeof(WCHAR) + 1 > *pdwSize)
+    if (!NT_SUCCESS(Status))
     {
-        Status = STATUS_BUFFER_TOO_SMALL;
         goto Cleanup;
     }
 
-    *pdwSize = Result->Length / sizeof(WCHAR);
-    memcpy(lpExeName, Result->Buffer, Result->Length);
-    lpExeName[*pdwSize] = 0;
+    /* The name should use the Win32 path format. */
+    if (dwFlags == 0)
+    {
+        pszResult = RtlAllocateHeap(RtlGetProcessHeap(), 0, Result->Length + sizeof(WCHAR));
```

Comment:
It may be possible to work, in this code path as well, without this redundant allocated buffer, and work with counted strings.